consensus/hotstuff/verification/common.go

```diff
@@ -66,3 +68,77 @@ func checkVotesValidity(votes []*model.Vote) error {
 
 	return nil
 }
+
+// stakingKeysAggregator is a structure that aggregates the staking
+// public keys for QC verifications.
+type stakingKeysAggregator struct {
+	lastStakingSigners map[*flow.Identity]struct{}
+	lastStakingKey     crypto.PublicKey
+	lock               sync.Mutex
+}
+
+// creates a new staking keys aggregator
+func newStakingKeysAggregator() *stakingKeysAggregator {
+	aggregator := &stakingKeysAggregator{
+		lastStakingSigners: map[*flow.Identity]struct{}{},
+		lastStakingKey:     crypto.NeutralBLSPublicKey(),
+		lock:               sync.Mutex{},
+	}
+	return aggregator
+}
+
+// aggregatedStakingKey returns the aggregated public of the input signers.
+func (s *stakingKeysAggregator) aggregatedStakingKey(signers flow.IdentityList) (crypto.PublicKey, error) {
+	s.lock.Lock()
```

Comment:
if we lock the entire `aggregatedStakingKey` function, then it might behind the bottleneck. 

The lock is only needed when reading and writing the `lastStakingSigners` and `lastStakingKey`, and the aggregating could be done in concurrently:


```
func (s *stakingKeysAggregator) aggregatedStakingKey(signers flow.IdentityList) (crypto.PublicKey, error) {
  var lastStakingSigners ...
  var lastStakingKeys ...
  s.Lock()
  lastStakingSigners = s.lastStakingSigners
  lastStakingKeys = s.lastStakingKeys
  s.UnLock()

  newLastStakingSigners, newLastStakingKeys, err = aggregate(signers, lastStakingSigners, lastStakingKeys)
  if err != nil {
    return nil, err
  }

  s.Lock()
  s.lastStakingSigners = newLastStakingSigners
  s.lastStakingKeys = newLastStakingKeys
  s.UnLock()
}
```

Reply:
You're right, good point.
What do you think of:
 - using [RWMutex](https://golang.org/pkg/sync/#RWMutex) instead of Mutex. This would block the parallel writing, and will only block the read when there is a writing happening. Parallel reads will be allowed. 
 - I can also make the writing done in a separate thread to unblock the function.

Reply:
I've updated the code with `1.`, but I could also add `2.`  at the end of the function:
```
 func(..) (..) {
         ...
         ...
	// update the latest list and public key. 
	go func(.., .. ) {
	   s.Lock()
	   s.lastStakingSigners = updatedList
	   s.lastStakingKey = updatedKey
	   s.Unlock()
	}(.., ..)
	return updatedKey, nil
```

Reply:
I like the current implementation with 
https://github.com/onflow/flow-go/blob/bd883091e558748a6d9364cfbac07f0bd0e0ebef/consensus/hotstuff/verification/common.go#L102-L105

https://github.com/onflow/flow-go/blob/bd883091e558748a6d9364cfbac07f0bd0e0ebef/consensus/hotstuff/verification/common.go#L124-L127